### PR TITLE
Extrapolation gp

### PR DESCRIPTION
**Pull request by garypolzin**
Extrapolate values before/after first/last known job_counts and input as a step into estimate_job_count process

[Trello ticket](https://trello.com/c/5idvj8YB/62-extrapolation-model-to-check)

### CODE REVIEW

File: tests/unit/test_extrapolation.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            generate_data_for_creating_extrapolated_ratios_df()
+        )
+        self.data_for_extrapolated_model_outputs_df = (
+            generate_data_for_creating_extrapolated_model_outputs_df()
+        )
+
+        warnings.filterwarnings("ignore", category=ResourceWarning)
+        warnings.filterwarnings("ignore", category=DeprecationWarning)
+
+    def test_model_extrapolation_row_count_unchanged(self):
```
Comment (MarionJHolloway) [inline tests/unit/test_extrapolation.py:42]: As with the interpolation, it might be worth adding an integrated test of the whole of model_extrapolation()

**Review by MarionJHolloway — COMMENTED** [review]
Looking good - nearly there!

File: utils/estimate_job_count/models/extrapolation.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    return left_join_on_locationid(max_job_count_per_location_df, df)
+
+
+def calculate_max_job_count_for_each_location(
+    df: pyspark.sql.DataFrame,
+) -> pyspark.sql.DataFrame:
+    return df.groupBy(LOCATION_ID).agg(F.max(JOB_COUNT).alias(MAX_JOB_COUNT))
+
+
+def filter_to_known_values_only(
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/estimate_job_count/models/extrapolation.py:65]: yea this was intentional - it should never have actual 0 staff if it's an open location. Even if it's dormant someone will still need to be in charge

File: utils/estimate_job_count/models/extrapolation.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    return df.join(
+        df_with_extrapolation_models, [LOCATION_ID, SNAPSHOT_DATE], "leftouter"
+    )
+
+
+def create_extrapolation_ratio_column(
+    df: pyspark.sql.DataFrame,
+) -> pyspark.sql.DataFrame:
+    return df.withColumn(
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/estimate_job_count/models/extrapolation.py:148]: yep, think I've done that now

File: tests/unit/test_extrapolation.py (hunk 1 above)
Comment (garypolzin) [reply tests/unit/test_extrapolation.py:42]: done :)

File: utils/estimate_job_count/models/extrapolation.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+LAST_ROLLING_AVERAGE = "last_rolling_average"
+LAST_JOB_COUNT = "last_job_count"
+EXTRAPOLATION_RATIO = "extrapolation_ratio"
+EXTRAPOLATION_MODEL = "extrapolation_model"
+
+
+def model_extrapolation(df: DataFrame) -> DataFrame:
+    for_extrapolation = filter_to_locations_who_have_a_job_count_at_some_point(df)
+
+    for_extrapolation = add_job_count_and_rolling_average_for_first_and_last_submission(
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/estimate_job_count/models/extrapolation.py:31]: renamed files to hopefully be more descriptive

File: utils/estimate_job_count/models/extrapolation.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        df, filtered_with_first_and_last_submitted_data_df
+    )
+
+    df_with_extrapolated_values = df_with_extrapolated_values.withColumn(
+        ESTIMATE_JOB_COUNT,
+        F.when(
+            F.col(ESTIMATE_JOB_COUNT).isNotNull(), F.col(ESTIMATE_JOB_COUNT)
+        ).otherwise(F.col(EXTRAPOLATION_MODEL)),
+    )
+    return update_dataframe_with_identifying_rule(
```
Comment (MarionJHolloway) [inline utils/estimate_job_count/models/extrapolation.py:45]: Can we separate out the return statement here? It's not immediately clear what the "update dataframe..." function returns, so it masks what model_extrapolation is returning.

File: utils/estimate_job_count/models/extrapolation.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        ).when(
+            (F.col(UNIX_TIME) > F.col(LAST_SUBMISSION_TIME)),
+            (
+                1
+                + (F.col(ROLLING_AVERAGE_MODEL) - F.col(LAST_ROLLING_AVERAGE))
+                / F.col(LAST_ROLLING_AVERAGE)
+            ),
+        ),
+    )
+    return df_with_extrapolation_ratio_column_df
```
Comment (MarionJHolloway) [inline utils/estimate_job_count/models/extrapolation.py:171]: This variable name has df in it twice. Can probably remove one of them

**Review by MarionJHolloway — CHANGES_REQUESTED** [review]
Looking really good, just a couple of small tweaks I spotted. Be careful when you fold functions into return statements that you don't obfuscate what datatype(s) you are actually returning! It helps to be able to understand the inputs and outputs of function definitions at a glance when you come to edit the code later down the line.

File: utils/estimate_job_count/models/extrapolation.py (hunk 6 above)
Comment (garypolzin) [reply utils/estimate_job_count/models/extrapolation.py:171]: sorted

File: utils/estimate_job_count/models/extrapolation.py (hunk 5 above)
Comment (garypolzin) [reply utils/estimate_job_count/models/extrapolation.py]: thanks, will do

**Review by MarionJHolloway — APPROVED** [review]
Looks good to me :)

**Pull request merged**